### PR TITLE
[linstor] Add ability to specify master passphrase

### PR DESCRIPTION
**Pull request by kvaps**
## Description

Add ability to specify master passphrase for LINSTOR

## Why do we need it, and what problem does it solve?

Some features requires to specify master passphrase, eg https://github.com/deckhouse/deckhouse/pull/1966

## What is the expected result?

User can create secret `d8-system/linstor-passphrase`, it should be copied into `d8-linstor/linstor-passphrase` and enabled in `luksSecret` field in linstorController resource.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: linstor
type: feature
summary: Add ability to specify master passphrase
impact: Master password enables some features like backup shipping and volume encryption using LUKS.
impact_level: high
```

### CODE REVIEW

File: modules/031-linstor/docs/CONFIGURATION.md [hunk 1]
```diff
@@ -9,10 +9,24 @@ data:
   linstorEnabled: "true"
 ```
 
-The module requires no configuration and has no parameters.
-
 After enabling the module, the cluster is automatically configured to use LINSTOR, and all that remains is to configure the storage.
 
+The module requires no configuration and has no parameters. However, some functions may require a master passphrase.  
+To set a master passphrase, create a secret in the `d8-system` namespace:
```
Comment (z9r5) [inline modules/031-linstor/docs/CONFIGURATION.md:15]:
```suggestion
To set a master passphrase, create a Secret in the `d8-system` namespace:
```

File: modules/031-linstor/docs/CONFIGURATION_RU.md [hunk 2]
```diff
@@ -9,10 +9,24 @@ data:
...
+kind: Secret
+metadata:
+  name: linstor-passphrase
+  namespace: d8-system
+immutable: true
+stringData:
+  MASTER_PASSPHRASE: *!пароль* # Мастер-пароль для LINSTOR
+```
+
+> **Внимание**: подойдите ответсвенно к выбору мастер-пароля для LINSTOR. При его потере зашифрованные данные окажутся недоступными
```
Comment (z9r5) [inline modules/031-linstor/docs/CONFIGURATION_RU.md:28]:
```suggestion
> **Внимание**: подойдите ответственно к выбору мастер-пароля для LINSTOR. При его потере зашифрованные данные окажутся недоступными.
```

**Review by z9r5 — APPROVED** [review]

File: modules/031-linstor/docs/CONFIGURATION.md [hunk 3]
```diff
@@ -9,10 +9,24 @@ data:
...
+kind: Secret
+metadata:
+  name: linstor-passphrase
+  namespace: d8-system
+immutable: true
+stringData:
+  MASTER_PASSPHRASE: *!passphrase* # Master passphrase for LINSTOR
+```
+
+> **Warning**: Be responsible to choose the master passphrase for LINSTOR. If it get lost, the encrypted data will be inaccessible.
```
Comment (shvgn) [inline modules/031-linstor/docs/CONFIGURATION.md:28]:
```suggestion
> **Warning**: Choose strong passphrase and store it securely. If it get lost, the encrypted data will be inaccessible.
```

Comment (shvgn) [reply modules/031-linstor/docs/CONFIGURATION.md:28]: BTW `password` seem to be a little more common term that `passphrase`. I'd probably stick to `password`.

File: modules/031-linstor/hooks/read_passphrase.go [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+	var passphrase string
+
+	snaps := input.Snapshots["master_passphrase"]
+	for _, snap := range snaps {
+		passphrase = snap.(string)
+	}
+
+	if passphrase != "" {
+		input.Values.Set("linstor.internal.masterPassphrase", passphrase)
+	}
```
Comment (shvgn) [inline modules/031-linstor/hooks/read_passphrase.go:74]: What if the passphrase has been set once and then cleaned (either changed to empty string, or the secret has been deleted). Should we clean the value?

**Review by shvgn — COMMENTED** [review]
Questioned opened

File: modules/031-linstor/hooks/read_passphrase.go (hunk 4 above)
Comment (yalosev) [reply modules/031-linstor/hooks/read_passphrase.go:74]: Probably, you don't need ExecuteOnEvent here

File: modules/031-linstor/docs/CONFIGURATION.md (hunk 3 above)
Comment (kvaps) [reply modules/031-linstor/docs/CONFIGURATION.md:28]: This is [official term of linstor](https://linbit.com/drbd-user-guide/linstor-guide-1_0-en/#s-linstor-encrypted-volumes), I would not change it.

File: modules/031-linstor/hooks/read_passphrase.go (hunk 4 above)
Comment (kvaps) [reply modules/031-linstor/hooks/read_passphrase.go:76]: > What if the passphrase has been set once and then cleaned (either changed to empty string, or the secret has been deleted). Should we clean the value?

Nice catch, thank you!

> Probably, you don't need ExecuteOnEvent here

I didn't get this, could you clarify please?

Comment (yalosev) [reply modules/031-linstor/hooks/read_passphrase.go:76]: You have `OnBeforeHelm` binding and I assume, you don't need to watch changes in the secret (do you expect it to be changed by external, don't you?)
Thats why you can add `ExecuteHookOnEvents: false` to the kubernetes binding

**Review by shvgn — APPROVED** [review]

File: modules/031-linstor/hooks/read_passphrase.go (hunk 4 above)
Comment (kvaps) [reply modules/031-linstor/hooks/read_passphrase.go:78]: Got it, thanks!

Comment (shvgn) [reply modules/031-linstor/hooks/read_passphrase.go:78]: For deleted secret or cleaned key from it, we should have a test case.

**Review by shvgn — CHANGES_REQUESTED** [review]

File: modules/031-linstor/hooks/read_passphrase.go (hunk 4 above)
Comment (kvaps) [reply modules/031-linstor/hooks/read_passphrase.go:78]: We have all of them:

https://github.com/deckhouse/deckhouse/blob/5a635b374875dca3eef4555278a4c3613cd30303/modules/031-linstor/hooks/read_passphrase_test.go#L29-L86

Comment (shvgn) [reply modules/031-linstor/hooks/read_passphrase.go:78]: You show me the case when the initial state is empty. What I propose, it to check the flow of creation and deletion of the passphrase. And check that after it was created and then cleaned, the value is not stuck

**Review by yalosev — APPROVED** [review]

File: modules/031-linstor/hooks/read_passphrase.go (hunk 4 above)
Comment (kvaps) [reply modules/031-linstor/hooks/read_passphrase.go:78]: Got it! Test case added
https://github.com/deckhouse/deckhouse/blob/2ce3ba7bc9f3926484377751c7fc614d0da8bbcf/modules/031-linstor/hooks/read_passphrase_test.go#L111-L122

**Review by shvgn — APPROVED** [review]

**Review by konstantin-axenov — APPROVED** [review]

**Pull request merged**